### PR TITLE
NativeAot-LLVM: Merge before putarg type removal PR 68748

### PR DESCRIPTION
**Pull request by yowl**
This PR merges up the commit before https://github.com/dotnet/runtime/pull/68748 which I'll do on it's own.  

Of note:

Added `PreAllocatedOverlapped.NativeAot.Wasm.cs`  under `nativeaot` as that is now excluded in `src/libraries`.  I hope when we enable wasm threads we can just delete this and say pthreads are not supported, which will bring in the file in `src/libraries` and make this redundant.

We changed to .net 7

### CODE REVIEW

**Comment by SingleAccretion** [comment]
@yowl (I know this is fraft, but) is [ffe71b9](https://github.com/dotnet/runtimelab/pull/2187/commits/ffe71b9a4a929b4f8b24624d0f2f586bc4b2350f) (Apr 26) the right base commit? The last merged one was [b39c723](https://github.com/dotnet/runtimelab/pull/2181/commits/b39c723a9b6c07f85304bf41a97bcc18225b1206) from Apr 20.

**Comment by yowl** [comment]
> @yowl (I know this is fraft, but) is [ffe71b9](https://github.com/dotnet/runtimelab/pull/2187/commits/ffe71b9a4a929b4f8b24624d0f2f586bc4b2350f) (Apr 26) the right base commit? The last merged one was [b39c723](https://github.com/dotnet/runtimelab/pull/2181/commits/b39c723a9b6c07f85304bf41a97bcc18225b1206) from Apr 20.

Ug, how did I manage that I wonder.

**Comment by yowl** [comment]
> > @yowl (I know this is fraft, but) is [ffe71b9](https://github.com/dotnet/runtimelab/pull/2187/commits/ffe71b9a4a929b4f8b24624d0f2f586bc4b2350f) (Apr 26) the right base commit? The last merged one was [b39c723](https://github.com/dotnet/runtimelab/pull/2181/commits/b39c723a9b6c07f85304bf41a97bcc18225b1206) from Apr 20.

Is that message reliable I wonder - how does github know after the merge is done?  I ran 
```
E:\GitHub\runtimelab\src>git merge-base b7a5420277efb00a4d0d5ffddce510e0cde1d65e 27f6e08178346b61ec49fc64f176022079530661
b39c723a9b6c07f85304bf41a97bcc18225b1206
```

And the base `b39c723a9b6c07f85304bf41a97bcc18225b1206` is correct.  

`b7a5420277efb00a4d0d5ffddce510e0cde1d65e ` is our tip in runtimelab/feature/NativeAOT-LLVM
`27f6e08178346b61ec49fc64f176022079530661` is the last merged commit here.

I then checked the file in the commit after `1b329a0ec6d354f3cc6cd4b699fff6209d9e4858`   https://github.com/dotnet/runtime/commit/1b329a0ec6d354f3cc6cd4b699fff6209d9e4858

And that change is here, 

https://github.com/yowl/runtimelab/blob/6ca30d7fcb796aae5148df484d88021faf15c85d/src/libraries/System.Diagnostics.Process/tests/ProcessModuleTests.cs#L4

From : 
https://github.com/dotnet/runtime/pull/68104/files

This page does not look coherent on the commit numbers either, it says 401 commits, but then in the list there is 30 shown, 192 hidden, then 27 more shown which clearly is not 401.  Happy to try other merge commands, but so far I can't make it do anything different as I understand the base is calculated automatically.

**Comment by SingleAccretion** [comment]
Good point, looks like GH is just not able to work things out here. I fetched the branch and it looks good based on `git log --graph --oneline FETCH_HEAD`:
```
* 6ca30d7fcb7 fix up compilation errors
* 94b442003a9 resolve conflicts
*   4f8db16fc6b Merge commit '27f6e08178346b61ec49fc64f176022079530661' into merge-before-putarg-type-removal
|\
| * 27f6e081783 Clean up specialized value num functions (#69013)
| *   ff237352b2c Merge pull request #67771 from dotnet/darc-main-31d4c308-6782-4440-8589-61f58f239059
| |\
| | * 9b73f24d519 Update dependencies from https://github.com/dotnet/arcade build 20220505.2
| | * 60593e4e430 Revert bad version reversions from 85d2aff4da96ac01babd461a1c76c358d4979d68
| | * b29d8c70291 Update dependencies from https://github.com/dotnet/linker build 20220506.1
| | *   35f0f338cfe Merge remote-tracking branch 'upstream/main' into darc-main-31d4c308-6782-4440-8589-61f58f239059
| | |\
| | * \   cfbf3035060 Merge remote-tracking branch 'upstream/main' into darc-main-31d4c308-6782-4440-8589-61f58f239059
| | |\ \
| | * \ \   3730e89ba2b Merge remote-tracking branch 'upstream/darc-main-f4af19cb-ce6b-425c-9c5b-dbc3910ea794' into darc-main-31d4c308-6782-4440-8589-61f58f239059
| | |\ \ \
| | | * | | ccf828380ce Update dependencies from https://github.com/dotnet/linker build 20220505.1
| | | * | | 990c92e91f5 Update dependencies from https://github.com/dotnet/linker build 20220503.2
| | * | | | 13e53a65509 Fix ProfileOptimization test
| | * | | | 02e68be226a Comment out assert for now
| | * | | |   e8264d8c7a2 Merge remote-tracking branch 'upstream/main' into darc-main-31d4c308-6782-4440-8589-61f58f239059
| | |\ \ \ \
| | * | | | | ce7dce9bd2c Fix ApiCompat errors
| | * | | | | 4316f371668 Update dependencies from https://github.com/dotnet/arcade build 20220504.2
| | * | | | |   be4f8fa79c2 Merge remote-tracking branch 'upstream/main' into darc-main-31d4c308-6782-4440-8589-61f58f239059
| | |\ \ \ \ \
| | * | | | | | 604a02348ed Comment out WASM vsman - manual process
| | * | | | | | fcc2cd478eb Add token to FF code
| | * | | | | |   85d2aff4da9 Merge branch 'main' into darc-main-31d4c308-6782-4440-8589-61f58f239059
| | |\ \ \ \ \ \
| | * | | | | | | bc1cffa6312 Pass runtime identifier on command line to work around dotnet/sdk#25062
| | * | | | | | | 844fd3db6a2 Update dependencies from https://github.com/dotnet/arcade build 20220425.6
| | * | | | | | |   4b941d3f131 Merge remote-tracking branch 'origin/main' into darc-main-31d4c308-6782-4440-8589-61f58f239059
| | |\ \ \ \ \ \ \
| | * | | | | | | | e596ee606d9 Update dependencies from https://github.com/dotnet/arcade build 20220422.4
| | * | | | | | | | a11ad178688 Update dependencies from https://github.com/dotnet/arcade build 20220421.2
| | * | | | | | | | 503a15de848 Revert temporary XHarness changes from #68024
| | * | | | | | | |   4b18acfe866 Merge remote-tracking branch 'dotnet/main' into darc-main-31d4c308-6782-4440-8589-61f58f239059
| | |\ \ \ \ \ \ \ \
| | * \ \ \ \ \ \ \ \   fe5c82b1e9e Merge branch 'main' into darc-main-31d4c308-6782-4440-8589-61f58f239059
| | |\ \ \ \ \ \ \ \ \
| | * | | | | | | | | | 67b548a34f2 Update dependencies from https://github.com/dotnet/arcade build 20220420.6
| | * | | | | | | | | | d19cc36998f Try using the centos 8 package to see if Ruby version is the problem
| | * | | | | | | | | | fe3c00b4be7 Update dependencies from https://github.com/dotnet/arcade build 20220418.3
| | * | | | | | | | | | 362770a5fa1 Update IJW.cmake
| | * | | | | | | | | | ba1778eb52f Update dependencies from https://github.com/dotnet/arcade build 20220415.2

...

| * | | | | | | d47140933dd Added `TestEventListener` (#68273)
| | |_|_|_|_|/
| |/| | | | |
| * | | | | | d58efd1f78c Add more details to NativeAOT workflow doc (#68308)
| * | | | | | 2894f5422c0 Update Microsoft.DotNet.Helix.Sdk (#68024)
| * | | | | | a4d45b2c8ed Add more diagnostics to LookupProperty (more diagnostics to flaky tests) (#68281)
| * | | | | | c78bf2f522b Update TestHandleCollector() test (#68285)
| * | | | | | 6ec5931a2df Add API entry for JsonDocument converter (#60236)
| * | | | | | d5ce000205e Use RegexCaseEquivalence table for case-insensitive backreferences (#67977)
| * | | | | | 9594b9c9ad8 Remove dead const from RegexNode (#68297)
| * | | | | | 1b329a0ec6d Add retries to ModulesAreDisposedWhenProcessIsDisposed (#68104)
* | | | | | |   b7a5420277e (HEAD -> feature/NativeAOT-LLVM, upstream/feature/NativeAOT-LLVM) Merge pull request #2181 from yowl/merge-apr-20
|\ \ \ \ \ \ \
| * \ \ \ \ \ \   ecd1d407873 Merge branch 'merge-apr-20' of https://github.com/yowl/runtimelab into merge-apr-20
| |\ \ \ \ \ \ \
| | * | | | | | | ec35fe0d8e8 Update liveBuilds.targets
| * | | | | | | | ff85da87011 Feedback: change Dump to match upstream
| |/ / / / / / /
| * | | | | | | 706ea11c24d try excluding ResolveRuntimeFilesFromLocalBuild for all nativeaot test builds
| * | | | | | | 265e37696f0 remove ArgNum
| * | | | | | | 65ab19c4a6d ignore lack of RuntimeFiles as there are none.
| * | | | | | | 91b43947113 resolve conflicts and allow nullable assignment
| * | | | | | |   7f6cddf96c8 Merge remote-tracking branch 'origin/feature/NativeAOT-LLVM' into clrjit-args
| |\ \ \ \ \ \ \
| * \ \ \ \ \ \ \   56fc3b85a5f Merge commit 'b39c723a9b6c07f85304bf41a97bcc18225b1206' into clrjit-args
| |\ \ \ \ \ \ \ \
| | | |/ / / / / /
| | |/| | | | | |
| | * | | | | | | b39c723a9b6 Fix Docs_Examples_ValidateEmail test (#68294)
```
As well as manually `git diff`ing the merge commit against its upstream counterpart.

File: src/coreclr/tools/Common/JitInterface/CorInfoBase.cs
Comment (SingleAccretion) [inline src/coreclr/tools/Common/JitInterface/CorInfoBase.cs]: Hmm, this will break the next time someone updates Jit-EE interface, since the file is auto-generated.

So we either need to update the thunk generator or add `InternalsVisibleTo` for the assembly which needs access, which I think is a better option (fewer conflicts).

File: src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs
Comment (SingleAccretion) [inline src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs]: With `InternalsVisibleTo`, it should be possible to undo the `public` and synchronize with upstream.

File: src/coreclr/tools/aot/ILCompiler.RyuJit/Compiler/RyuJitCompilationBuilder.cs
Comment (SingleAccretion) [inline src/coreclr/tools/aot/ILCompiler.RyuJit/Compiler/RyuJitCompilationBuilder.cs]:
```suggestion
            : this(context, group, context.Target.IsWindows ? new WindowsNodeMangler() : new UnixNodeMangler())
```
Let's resolve this in our favor since it has simpler code.

File: src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.Llvm.cs
Comment (SingleAccretion) [inline src/coreclr/tools/aot/ILCompiler.RyuJit/JitInterface/CorInfoImpl.Llvm.cs]:
```suggestion
    public unsafe sealed partial class CorInfoImpl
```

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets
Comment (SingleAccretion) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets]: I think it would be best to set `DisableUnsupportedError` to `true` for WASM somewhere above to avoid future conflicts in this section.

File: src/coreclr/tools/Common/JitInterface/CorInfoBase.cs
Comment (yowl) [reply src/coreclr/tools/Common/JitInterface/CorInfoBase.cs]: Added `InternalsVisibleTo`  had a knock effect with some duplicate classes which are now removed from the LLVM `csproj`

File: src/coreclr/tools/aot/ILCompiler.RyuJit/Compiler/RyuJitCompilationBuilder.cs
Comment (yowl) [reply src/coreclr/tools/aot/ILCompiler.RyuJit/Compiler/RyuJitCompilationBuilder.cs]: Done, thanks

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets
Comment (yowl) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets]: Added near the top, thanks

**Comment by yowl** [comment]
Sorry, my .net 7 installation is not letting me run the tests locally using xunit at the moment.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets
Comment (SingleAccretion) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets]:
```suggestion
    <Error Condition="'$(DisableUnsupportedError)' != 'true' and '$(OS)' == 'Windows_NT' and !$(RuntimeIdentifier.StartsWith('win'))"
```

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets
Comment (SingleAccretion) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets]:
```suggestion
    <Error Condition="'$(DisableUnsupportedError)' != 'true' and '$(OS)' != 'Windows_NT' and $(RuntimeIdentifier.StartsWith('win'))"
```

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets
Comment (SingleAccretion) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Publish.targets]:
```suggestion
      ($(RuntimeIdentifier.EndsWith('arm64')) and !$(RuntimeIdentifier.StartsWith('osx'))))"
```

**Review by SingleAccretion — APPROVED** [review]
LGTM (once CI is green)!

**Comment by jkotas** [comment]
Thanks!

**Pull request merged**
